packages/Search/src/details/DTK_DetailsDistributor.hpp

```diff
@@ -0,0 +1,246 @@
+ * Copyright (c) 2012-2018 by the DataTransferKit authors                   *
+ * All rights reserved.                                                     *
+ *                                                                          *
+ * This file is part of the DataTransferKit library. DataTransferKit is     *
+ * distributed under a BSD 3-clause license. For the licensing terms see    *
+ * the LICENSE file in the top-level directory.                             *
+ *                                                                          *
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+#ifndef DTK_DETAILS_DISTRIBUTOR_HPP
+#define DTK_DETAILS_DISTRIBUTOR_HPP
+
+#include <Kokkos_Core.hpp> // FIXME
+
+#include <mpi.h>
+
+#include <algorithm>
+#include <cassert>
+#include <vector>
+
+#define REORDER_RECV YES
+
+namespace DataTransferKit
+{
+namespace Details
+{
+
+// clang-format off
+// copy/paste from https://github.com/kokkos/kokkos-tutorials/blob/ee619447cee9e57b831e34b151d0868958b6b1e5/Intro-Full/Exercises/mpi_exch/mpi_exch.cpp
+// * added Experimental nested namespace qualifier for Max
+// * declared function static to avoid multiple definition error at link time
+// * return early if input argument is empty
+// * replace parallel_reduce with std::max_element because I couldn't get it to compile
+// * append size to offsets in main driver but might want to handle it here
+static void extract_and_sort_ranks(
+    Kokkos::View<int*, Kokkos::HostSpace> destination_ranks,
+    Kokkos::View<int*, Kokkos::HostSpace> permutation,
+    std::vector<int>& unique_ranks,
+    std::vector<int>& offsets,
+    std::vector<int>& counts) {
+  int mpi_rank;
+  MPI_Comm_rank(MPI_COMM_WORLD, &mpi_rank);
+  auto n = destination_ranks.extent(0);
+if ( n == 0 ) return;
+  using ST = decltype(n);
+  Kokkos::View<int*, Kokkos::HostSpace> tmp_ranks("tmp ranks", destination_ranks.extent(0));
+  Kokkos::deep_copy(tmp_ranks, destination_ranks);
+  int offset = 0;
+  // this implements a "sort" which is O(N * R) where (R) is
+  // the total number of unique destination ranks.
+  // it performs better than other algorithms in
+  // the case when (R) is small, but results may vary
+  while (true) {
+    int next_biggest_rank;
+    //Kokkos::parallel_reduce("find next biggest rank", n, KOKKOS_LAMBDA(ST i, int& local_max) {
+    //  auto r = tmp_ranks(i);
+    //  local_max = (r > local_max) ? r : local_max;
+    //}, Kokkos::Experimental::Max<int, Kokkos::HostSpace>(next_biggest_rank));
+    next_biggest_rank = *std::max_element(tmp_ranks.data(), tmp_ranks.data() + n);
+    if (next_biggest_rank == -1) break;
+    unique_ranks.push_back(next_biggest_rank);
+    offsets.push_back(offset);
+    Kokkos::View<int, Kokkos::HostSpace> total("total");
+    Kokkos::parallel_scan("process biggest rank items", Kokkos::RangePolicy<Kokkos::Serial>(0, n),
```

Comment:
I am confused are you doing the sort in serial? Then why not use STL

Reply:
Copy/paste from Kokkos tutorial.  Will revisit.